### ISSUE TITLE
ACT rule "Visible label is part of accessible name" (2ee8b8) changes break nightly

### ISSUE THREAD

**Issue by straker** [dequelabs/axe-core#5207]
Latest ACT update for our rule 2ee8b8 "Visible label is part of accessible name" has [added more examples](https://github.com/w3c/wcag-act-rules/blob/main/content/rules/2ee8b8/proposed.md) which are [breaking our nightly tests run](https://github.com/dequelabs/axe-core/actions/runs/29303327772/job/86991534641). The changes are non-trival as they follow a [new algorithm](https://github.com/w3c/wcag-act-rules/blob/main/content/rules/2ee8b8/proposed.md#label-in-name-algorithm-label-in-name-algorithm) which we don't implement. So things like removing all content between round brackets and maintaining space between spans are currently failing. [Our code currently uses visibleVirtual](https://github.com/dequelabs/axe-core/blob/develop/lib/checks/label/label-content-name-mismatch-evaluate.js#L45-L49) to generate the visible content, which we'll need to figure out if we want to continue to use that function or create a new function just for this algorithm.

This issue is linked to a `skipTests` entry in `test/act-rules/visible-label-in-accessible-name-2ee8b8.spec.js`.

**Comment by chutchins25** [comment]
Picking this up. Step 1 (parenthetical removal) is implemented in #5360 — `Passed Example 14` now passes unskipped, at 35 passing / 3 pending / 0 failing against `wcag-act-rules#main`.

The `visibleVirtual` question you raised is the one that needs your call, and it's sharper than it looks. `Passed Example 11` is this:

```html
<a aria-label="Download specification"
  ><span>Download</span><span id="space"> </span><span>specification</span></a
>
```

`visibleVirtual` recurses per child and ends each recursive call with `sanitize()`, which trims. So the whitespace-only `<span> </span>` returns `''`, the `.join('')` produces `downloadspecification`, and no amount of tokenising recovers the word boundary. It's `lib/commons/text/visible-virtual.js:63`, not the algorithm.

That leaves two options:

1. **Fix `visibleVirtual`.** Correct, but it feeds `color-contrast`, `target-size` via `create-grid`, modal detection and several matchers, so it changes what "visible text" means for every rule that reads it — the same blast radius that got #3883 de-scoped from this epic.
2. **A rule-local text function.** Contained, but it duplicates flattened-tree traversal and means `label-content-name-mismatch` stops agreeing with the rest of axe about what text is visible.

I'd lean towards 1, as its own PR with a deliberate regression pass rather than folded into rule work — but it's your rule and your call.

**Comment by chutchins25** [comment]
Both halves are on `develop`:

- Step 1 of the label in name algorithm — removing parenthesised content — in #5360 (`388132db`).
- The `visibleVirtual` whitespace defect, fixed as "sanitize only at the top of the recursion" per the design call on this issue, in #5368 (`2d471821`).

ACT `2ee8b8` against `wcag-act-rules#main` is now **36 passing / 2 pending / 0 failing**. `Passed Example 11`'s `skipTests` entry is removed; the two remaining pendings are the accepted abbreviation/hyphenation divergence (act-rules/act-rules.github.io#2443).

Closing manually — the `Closes issue #5207` footer on #5368 didn't fire because the PR targeted `develop` rather than the default branch.

**Comment by padmavemulapati** [comment]
Here's the comment — but run `qa.report()` first and adjust the counts line if anything comes back red; I don't want to hand you a sign-off for a run that hasn't happened.

```markdown
### QA validation — passed

Validated on `develop` @ `<commit>` (axe `<version>`) using a manual playground page covering both halves of this issue plus the word-wise containment change from #5302/#5345. 27 cases, all as expected.

**Parenthetical removal (#5360)**
- ACT Passed Example 14 passes unskipped.
- Nested bracket pairs are fully removed (the do/while loop, not just the outer pair).
- Removal preserves the surrounding word boundary — `Download (large) PDF` vs `Download PDF` passes.
- A label that is entirely parenthetical curates to zero tokens and is treated as contained (ab92f45).
- An unbalanced bracket is *not* stripped — `Close (x` vs `Close` still fails, as it should.

**visibleVirtual whitespace (#5368)**
- ACT Passed Example 11 passes; the whitespace-only `<span>` no longer collapses into `downloadspecification`.
- Holds one level deeper (whitespace-only span inside a nested span) and for `<br>`.
- Top-level `sanitize()` still collapses and trims runs of whitespace.
- Over-correction guard: `<span>Download</span><span> </span><span>specification</span>` against `aria-label="Downloadspecification"` still reports a violation, so the fix isn't just making everything pass.

**Containment, tokenisation, incompletes**
- Substring-but-not-word (`cat` vs "Concatenate values"), non-contiguous words and reversed word order all fail correctly; a contiguous run inside a longer name passes.
- Hyphen, emoji and soft hyphen (U+00AD) tokenise as expected — soft hyphen is stripped before punctuation handling, so it doesn't split the word.
- `valueText` incomplete fires for `role="switch"`; non-interpretable visible text returns the default incomplete.
- Applicability unchanged: no `aria-label`/`aria-labelledby`, non-widget roles, `textbox`, and elements with no visible text produce no result.

**One question before this is fully closed out.** axe runs `removeParentheticals` over the *accessible name* as well as the visible label, which ACT's algorithm doesn't do — it curates only the visible label. So `aria-label="Save (draft)"` with visible text `Save draft` reports a violation here, where ACT would pass it. It isn't covered by the current ACT examples so the nightly is green either way, but it's a behavioural difference worth a deliberate decision rather than an accident of implementation. Happy to raise it separately under #5226 if you'd rather not reopen this.

<img width="1719" height="985" alt="Image" src="https://github.com/user-attachments/assets/4299630c-1665-4c7d-b30d-72a38f747704" />
